coaster/views/misc.py

```diff
@@ -108,3 +111,49 @@ def jsonp(*args, **kw):
     else:
         mimetype = 'application/json'
     return Response(data, mimetype=mimetype)
+
+
+def endpoint_for(url, method=None, return_rule=False, follow_redirects=True):
+    """
+    Given an absolute URL, retrieve the matching endpoint name (or rule).
+    Requires a current request context to determine runtime environment.
+
+    :param str method: HTTP method to use (defaults to GET)
+    :param bool return_rule: Return the URL rule instead of the endpoint name
+    :param bool follow_redirects: Follow redirects to final endpoint
+    :return: Endpoint name or URL rule, or `None` if not found
+    """
+    parsed_url = urlsplit(url)
+    if not parsed_url.netloc:
+        # We require an absolute URL
+        return
+
+    # Take the current runtime environment...
+    environ = dict(request.environ)
+    # ...but replace the HTTP host with the URL's host...
+    environ['HTTP_HOST'] = parsed_url.netloc
+    # ...and the path with the URL's path (after discounting the app path, if not hosted at root).
+    environ['PATH_INFO'] = parsed_url.path[len(environ['SCRIPT_NAME']):]
```

Comment:
Do these need to be changed back to original value after this function is done with the forged request? Can this cause problem if any other piece of code called after calling `endpoint_for()` requires these environment variables?

Reply:
No, since we made a copy with `environ = dict(request.environ)`.

Reply:
Ah, right.